criu/pie/restorer.c

```diff
@@ -202,11 +202,31 @@ static int restore_creds(struct thread_creds_args *args, int procfd,
 	 * Setup supplementary group IDs early.
 	 */
 	if (args->groups) {
-		ret = sys_setgroups(ce->n_groups, args->groups);
-		if (ret) {
-			pr_err("Can't setup supplementary group IDs: %d\n", ret);
+		// We may be in an unprivileged user namespace where setgroups
+		// is disabled.  If the current list of groups is already what
+		// we want, skip the call to setgroups.
+
+		int n = sys_getgroups(0, NULL);
+		int len = n * sizeof(unsigned int);
+		unsigned int* gids = (unsigned int*)sys_mmap(NULL, len, PROT_WRITE | PROT_READ,
+		       MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
+		if (gids == MAP_FAILED) {
+			pr_err("Unable to mmap %d bytes\n", len);
 			return -1;
 		}
+		n = sys_getgroups(n, gids);
+		if (n == ce->n_groups && !memcmp(gids, args->groups, len)) {
+			pr_debug("getgroups() looks good, not calling setgroups()\n");
+			ret = 0;
+		} else {
+			ret = sys_setgroups(ce->n_groups, args->groups);
+		}
+		if (ret)
+			pr_err("Can't setgroups([%zu gids]): %d\n", ce->n_groups, ret);
+		if (sys_munmap(gids, len))
+			pr_err("Unable to munmap %d bytes\n",len);
```

Comment:
The %d bytes part of this comment is unnecessary I think. The failure to unmap is unlikely to be due to size of the mapping.